src/libostree/ostree-repo-finder.h

```diff
@@ -0,0 +1,116 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#pragma once
+
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib-object.h>
+
+#include "ostree-types.h"
+
+G_BEGIN_DECLS
+
+#define OSTREE_TYPE_REPO_FINDER (ostree_repo_finder_get_type ())
+
+_OSTREE_PUBLIC
+G_DECLARE_INTERFACE (OstreeRepoFinder, ostree_repo_finder, OSTREE, REPO_FINDER, GObject)
+
+struct _OstreeRepoFinderInterface
+{
+  GTypeInterface g_iface;
+
+  void (*resolve_async) (OstreeRepoFinder    *self,
+                         const gchar * const *refs,
+                         GCancellable        *cancellable,
+                         GAsyncReadyCallback  callback,
+                         gpointer             user_data);
+  GPtrArray *(*resolve_finish) (OstreeRepoFinder  *self,
+                                GAsyncResult      *result,
+                                GError           **error);
+};
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_async (OstreeRepoFinder    *self,
+                                       const gchar * const *refs,
+                                       GCancellable        *cancellable,
+                                       GAsyncReadyCallback  callback,
+                                       gpointer             user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_finish (OstreeRepoFinder  *self,
+                                              GAsyncResult      *result,
+                                              GError           **error);
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_all_async (OstreeRepoFinder * const *finders,
+                                           const gchar * const      *refs,
+                                           GCancellable             *cancellable,
+                                           GAsyncReadyCallback       callback,
+                                           gpointer                  user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_all_finish (GAsyncResult  *result,
+                                                  GError       **error);
+
+/**
+ * OstreeRepoFinderResult:
+ * @uri: TODO
+ * @priority: TODO
+ *
+ * TODO: basically a structure which says ‘you can download these refs from
+ * this URI, and they will be up to date’, along with some relevant metadata.
+ *
+ * Since: 2017.6
+ */
+typedef struct
+{
+  /* TODO: Should I extend this to also support mirrors and metalink? Maybe add a GVariant *metadata too?
+   * For example, should a result be able to force GPG verification? */
+  /* TODO: would it support torrenting? */
+  gchar *uri;
```

Comment:
> The way I've been thinking of this is that the RepoFinder is basically a much fancier `mirrorlist` type implementation for _existing_ remotes. Basically, `flatpak update` would _try_ using repofinder, and if that was empty, would just fall back to its existing remote.

I think we’re still talking a bit cross-purposes here. What do you mean by ‘existing remote’? The Avahi stuff, by its very nature, cannot rely on ‘existing remotes’ since the set of peers on the local network can vary.

> It sounds like you're thinking of this more like "eos-updater does everything"? What would `flatpak update` do? Do you see it being ported to this?

Correct, I am thinking of this as effectively a replacement for the existing `ostree_repo_pull()` API.  `flatpak update` would _definitely_ be ported to this, and would pass it the refs for the set of flatpaks it wants to update (most likely, all the flatpaks the user has installed).

I hope I mentioned this before (I probably didn’t): the overall goal here is that flatpak can do updates from LAN and USB remotes, just like `eos-updater` can; and that the code bases for the two converge so that all the underlying LAN and USB functionality is in libostree. We probably eventually want to gut `eos-updater` and move it into a GNOME Software plugin, but that’s less relevant here.